query_optimizer/ExecutionGenerator.cpp

```diff
@@ -713,25 +795,73 @@ void ExecutionGenerator::convertHashJoin(const P::HashJoinPtr &physical_plan) {
-
-  // SimplifyHashTableImplTypeProto() switches the hash table implementation
-  // from SeparateChaining to SimpleScalarSeparateChaining when there is a
-  // single scalar key type with a reversible hash function.
-  hash_table_proto->set_hash_table_impl_type(
-      SimplifyHashTableImplTypeProto(
-          HashTableImplTypeProtoFromString(FLAGS_join_hashtable_type),
-          key_types));
+  const QueryContext::join_hash_table_group_id join_hash_table_group_index =
+      query_context_proto_->join_hash_table_groups_size();
+  query_context_proto_->add_join_hash_table_groups();
+  std::vector<S::HashTable *> hash_table_proto;
+
+  bool is_numa_aware_join = false;
+  if (build_relation_info->relation->hasPartitionScheme() && probe_operator_info->relation->hasPartitionScheme()) {
+    if (build_relation_info->relation->getPartitionScheme().getPartitionSchemeHeader().getNumPartitions() ==
+        probe_operator_info->relation->getPartitionScheme().getPartitionSchemeHeader().getNumPartitions()) {
+      // check if partitioning attribute == join attribute for both the probe and the build relations.
+      if (std::find(build_attribute_ids.begin(),
+                    build_attribute_ids.end(),
+                    build_relation_info->relation->getPartitionScheme()
+                        .getPartitionSchemeHeader()
+                        .getPartitionAttributeId()) != build_attribute_ids.end() &&
+          std::find(probe_attribute_ids.begin(),
+                    probe_attribute_ids.end(),
+                    probe_operator_info->relation->getPartitionScheme()
+                        .getPartitionSchemeHeader()
+                        .getPartitionAttributeId()) != probe_attribute_ids.end()) {
+        is_numa_aware_join = true;
+        const std::size_t num_partitions =
+            build_relation_info->relation->getPartitionScheme().getPartitionSchemeHeader().getNumPartitions();
+        hash_table_proto.resize(num_partitions);
+        for (std::size_t part_id = 0; part_id < num_partitions; ++part_id) {
+          hash_table_proto[part_id] = query_context_proto_->mutable_join_hash_table_groups(
+                                                                join_hash_table_group_index)->add_join_hash_tables();
+
+          // SimplifyHashTableImplTypeProto() switches the hash table implementation
+          // from SeparateChaining to SimpleScalarSeparateChaining when there is a
+          // single scalar key type with a reversible hash function.
+          hash_table_proto[part_id]->set_hash_table_impl_type(
+              SimplifyHashTableImplTypeProto(
+                  HashTableImplTypeProtoFromString(FLAGS_join_hashtable_type),
+                  key_types));
+
+          const CatalogRelationSchema *build_relation = build_relation_info->relation;
+          for (const attribute_id build_attribute : build_attribute_ids) {
+            hash_table_proto[part_id]->add_key_types()->CopyFrom(
```

Comment:
Can you please explain why since I don't understand why MergeFrom is preferred if both have the same functionality?